array.c

```diff
@@ -2043,30 +2073,54 @@ rb_ary_fetch(int argc, VALUE *argv, VALUE ary)
 
 /*
  *  call-seq:
- *     ary.find_index(obj)             ->  int or nil
- *     ary.find_index {|item| block}  ->  int or nil
- *     ary.find_index                  ->  Enumerator
- *     ary.index(obj)             ->  int or nil
- *     ary.index {|item| block}   ->  int or nil
- *     ary.index                  ->  Enumerator
+ *    array.index(object) -> integer or nil
+ *    array.index {|element| ... } -> integer or nil
+ *    array.index -> new_enumerator
+ *    array.find_index(object) -> integer or nil
+ *    array.find_index {|element| ... } -> integer or nil
+ *    array.find_index -> new_enumerator
  *
- *  Returns the _index_ of the first object in +ary+ such that the object is
- *  <code>==</code> to +obj+.
+ *  Array#index is an alias for Array#find_index.
  *
- *  If a block is given instead of an argument, returns the _index_ of the
- *  first object for which the block returns +true+.  Returns +nil+ if no
- *  match is found.
+ *  ---
  *
- *  See also Array#rindex.
+ *  When argument +object+ is given but no block,
+ *  returns the index of the first element +element+
+ *  for which <tt>object == element</tt>:
+ *    a = [:foo, 'bar', baz = 2, 'bar']
+ *    a.index('bar') # => 1
  *
- *  An Enumerator is returned if neither a block nor argument is given.
+ *  Returns +nil+ if no such element found:
+ *    a = [:foo, 'bar', baz = 2]
+ *    a.index(:nosuch) # => nil
  *
- *     a = [ "a", "b", "c" ]
- *     a.index("b")              #=> 1
- *     a.index("z")              #=> nil
- *     a.index {|x| x == "b"}    #=> 1
+ *  ---
  *
- *  Array#index is an alias for Array#find_index.
+ *  When both argument +object+ and a block are given,
+ *  calls the block with each successive element element;
+ *  returns the index of the first element for which the block returns a truthy value:
+ *    a = [:foo, 'bar', baz = 2, 'bar']
+ *    a.index { |element| element == 'bar' } # => 1
+ *
+ *  Returns +nil+ if the block never returns a truthy value:
+ *    a = [:foo, 'bar', baz = 2]
+ *    a.index { |element| element == :X } # => nil
+ *
+ *  ---
+ *
+ *  When neither an argument nor a block is given, returns a new Enumerator:
+ *    a = [:foo, 'bar', baz = 2]
+ *    e = a.index
+ *    e # => #<Enumerator: [:foo, "bar", 2]:index>
+ *    e.each { |element| element == 'bar' } # => 1
+ *
+ *  ---
+ *
+ *  When both an argument and a block given, gives a warning (warning: given block not used)
+ *  and ignores the block:
+ *    a = [:foo, 'bar', baz = 2, 'bar']
+ *    index = a.index('bar') { fail 'Cannot happen' }
```

Comment:
I would prefer to use `raise` instead of `fail` because most Rubyists use `raise` and don't know that they are aliases